### PR TITLE
Add binding generation for node

### PR DESCRIPTION
**Pull request by jt-nti**
Adds a node module template project and associate code generation

Signed-off-by: James Taylor <jamest@uk.ibm.com>

### CODE REVIEW

File: scripts/generate_node_indexes.sh [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+find_sub_dirs() {
+    for subDir in ${MODULE_DIRS}; do
+        echo "${subDir}" | grep "^$1/[^/]*$" | sed "s,$1/,,g"
+    done
+}
+
+pushd "${MODULE_SRC_DIR}" > /dev/null
+
+find . -type f \( -name '*.js' -o -name '*.d.ts' \) -exec grep -i 'GENERATED CODE -- NO SERVICES IN PROTO' {} \; -print -exec rm -f {} \;
```
Comment (bestbeforetoday) [inline scripts/generate_node_indexes.sh:21]: Not sure if `-delete` would be preferable to `-exec rm -f {} \;`

File: scripts/generate_node_indexes.sh [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+MODULE_FILES=$(find . -name '*.d.ts' -type f -print | sed 's/\.d\.ts$//g')
+MODULE_DIRS=$(find . -type d -print)
+
+for moduleDir in ${MODULE_DIRS}; do
+    pushd "${moduleDir}" > /dev/null
+
+    rm -rf index.ts
```
Comment (bestbeforetoday) [inline scripts/generate_node_indexes.sh:29]: I guess it doesn't hurt but this doesn't need the `-r` option as it's only supposed to be ensuring a single file doesn't exist -- my mistake

File: scripts/generate_node_indexes.sh (hunk 1 above)
Comment (jt-nti) [reply scripts/generate_node_indexes.sh]: New fangled :)

**Comment by jt-nti** [comment]
Closing due to DCO check not running

**Comment by jt-nti** [comment]
Thanks @mbwhite, I deliberately left publishing to a separate PR since it's likely to involve more than a little trial and error, and will require some dedicated support from someone with the right authority on the repo. Once this PR is in though I hope to start by publishing nightly dev versions to npmjs. Once the Java bindings are also in and publishing dev versions somewhere (Hyperledger artifactory?) we'll need to have a chat about versioning before publishing the first "proper" version of the new bindings. That's my vague plan anyway.

**Comment by mbwhite** [comment]
Sounds like a plan @jt-nti  - as it would become a transitive dependency of the published chaincode it needs to be readily accessible.... or more accurately as accessible as the chaincode libraries.

In the interim I guess I should clone this PR and produce a node module tgz file to test the API

**Comment by jt-nti** [comment]
@mbwhite that's what Mark did to test the new module in fabric-gateway... hyperledger/fabric-gateway#413

**Comment by jt-nti** [comment]
This PR is related to issue #72

**Comment by denyeart** [comment]
It is not clear from this PR what the proposed lifecycle is for proto and binding updates. Which parts will be automated, which parts manual, how the Makefile and scripts relate to each other, whether bindings get checked in to this repo or somewhere else, etc. The process seems very different from what is already in place for the Go bindings. Since consistency is a key motivation for this work item, we need to understand if the Go procedure will be modified to look more like the Node procedure or vice versa. This should be socialized in the original issue, with additional details going into a readme and/or commit message.

In terms of PR specifics, it is unclear which parts of the Makefile and scripts will actually be used in the proposed process versus just pulled in via copy/paste from somewhere else. For example Makefile references CircleCI which is obviously a copy/paste and not relevant for Fabric. We should keep the content to a minimum based on what will actually be used in the Fabric process instead of copying in extra cruft.

Finally, do we need to pollute the project root with the buf files? How will they be used?

We'll want @andrew-coleman to perform a final review as a key SDK stakeholder that consumes the protos, but these are my immediate questions.

**Comment by jt-nti** [comment]
Thanks @denyeart, I'll add a bit more detail to the issue.

> Finally, do we need to pollute the project root with the buf files? How will they be used?

They're comparable to the [prototool.yaml](https://github.com/hyperledger/fabric-protos/blob/main/prototool.yaml) file and are used to lint the protos and generate the language bindings.

**Review by andrew-coleman — APPROVED** [review]

**Pull request merged**